app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt

```diff
@@ -1,32 +1,97 @@
 package com.lightricks.feedexercise.ui.feed
 
+import android.util.Log
 import androidx.lifecycle.*
+import com.lightricks.feedexercise.R
 import com.lightricks.feedexercise.data.FeedItem
+import com.lightricks.feedexercise.data.FeedRepositoryImpl
+import com.lightricks.feedexercise.data.FeedRepository
+import com.lightricks.feedexercise.network.FeedApiService
 import com.lightricks.feedexercise.util.Event
-import java.lang.IllegalArgumentException
+import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
+import io.reactivex.rxjava3.disposables.Disposable
 
 /**
  * This view model manages the data for [FeedFragment].
  */
-open class FeedViewModel : ViewModel() {
-    private val isLoading = MutableLiveData<Boolean>()
-    private val isEmpty = MutableLiveData<Boolean>()
-    private val feedItems = MediatorLiveData<List<FeedItem>>()
+open class FeedViewModel(private val feedRepository: FeedRepository) : ViewModel() {
+
+    private val stateLiveData = MutableLiveData<State>()
     private val networkErrorEvent = MutableLiveData<Event<String>>()
+    private var prevTask: Disposable? = null
+    private lateinit var getFeedFromRepoStream: Disposable
+
+    fun getIsLoading(): LiveData<Boolean> = Transformations.map(stateLiveData) { it.isLoading }
+    fun getIsEmpty(): LiveData<Boolean> = Transformations.map(stateLiveData) { it.isFeedEmpty() }
+    fun getFeedItems(): LiveData<List<FeedItem>> =
+        Transformations.map(stateLiveData) { it.feedItems }
 
-    fun getIsLoading(): LiveData<Boolean> = isLoading
-    fun getIsEmpty(): LiveData<Boolean> = isEmpty
-    fun getFeedItems(): LiveData<List<FeedItem>> = feedItems
     fun getNetworkErrorEvent(): LiveData<Event<String>> = networkErrorEvent
 
     init {
+        observeRepository()
         refresh()
     }
 
     fun refresh() {
-        //todo: fix the implementation
-        isLoading.value = false
-        isEmpty.value = true
+        setStartLoading()
+        disposePreviousTask()
```

Comment:
I would first dispose the previous refresh task, and then setStartLoading. That is because I want first to cancel any influence the previous task could have on the view model. In this case there is no difference since the callback are invoked on the main thread as well. But as a best practice I think I would change the order.